Mannequins

JMWTurner opened the pull request:
I spent the entire day on this and it's 4:14 am. Screw this I'll write the changelog tomorrow.

![mannequins](https://cloud.githubusercontent.com/assets/21122521/18373387/f8a91dc6-7643-11e6-8d03-ecb5ff8d5cdf.png)
- They come in 3 varieties, wooden, marble, and cyber. Wooden and marble have fat and vox variants.
- You can dress them up
- Wooden mannequins are built by using a chisel on a wooden block, which is made from 10 wooden planks
- Marble mannequins are built by using a chisel on a marble block, which can be bought from cargo.
- Cyber mannequins frames are printed from an Exosuit Fabricator, you need to add robot parts for all limbs, an airlock circuit (which will determine what access can toggle the shield) and a glass sheet. It can then be toggled like a display case. It can also be broken and repaired.

DrCelt commented:
Can the cyber ones be highjacked by a malfunctioning AI?

CptWad commented:
while cool celt, lets do one thing at a time

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+		MANNEQUIN_ID_LAYER,
+		)
+
+	for(var/i = 1; i <= held_items.len; i++)
```
use a to loop here they're prettier

JMWTurner replied:
loop what? All those entries are of a different type.

RemieRichards replied:
a `for(var/i in 1 to held_items.len)`
it's faster too (because byond a shit)

JMWTurner replied:
oh, a "to" loop. Misread that.

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+	attack_hand(user)
+
+
+/obj/structure/mannequin/examine(mob/user)
```
this whole thing is copypasted, ew

JMWTurner replied:
copypasted and proof-read.

Intigracy replied:
Feel free to rewrite the entirety of inventorycode to be more modular if this is an issue, celt.

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+	update_icon_right_hand()
+
+
+/obj/structure/mannequin/Topic(href, href_list)
```
aAAAAAAAAAAAAA THIS IS ALL COPYPASTED

JMWTurner replied:
i-it's not! not all of it actually. I had to remove tons of stuff that don't apply to mannequins.

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+/obj/structure/mannequin/proc/get_index_limb_name(var/index)
+	switch(index)
+		if(GRASP_LEFT_HAND)
+			return "left hand"
```
Why

JMWTurner replied:
for both the examine() and show_inv() procs. Why else?

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+		if(anchored)
+			to_chat(user, "<span class='info'>You start unsecuring \the [src] from \the [loc].</span>")
+		else
+			if(!istype(loc, /turf/simulated/floor))
```
So unsimulated turfs can't be wrenched on?

ComicIronic commented on code/game/objects/structures/displaycase.dm:
```diff
@@ -138,7 +138,8 @@
 
 /obj/structure/displaycase/bullet_act(var/obj/item/projectile/Proj)
-	health -= Proj.damage
+	if(!istype(Proj ,/obj/item/projectile/beam/lasertag) && !istype(Proj ,/obj/item/projectile/beam/practice) )
```
Non-atomic.

Exxion replied:
1. What the FUCK is up with your spaces?
2. Don't those projectiles deal 0 damage anyway? At least they should.
3. I'm 90% sure projectiles such as these two have a variable specifying they should never deal damage. Use that rather than a hardcoded type check.

JMWTurner replied:
They used to have actual damage.

ComicIronic commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+		if (3)
+			if (prob(50))
+				health -= 10
+				healthcheck()
```
Rather than requiring a healthcheck proc separately, it would be saner to have a proc for changing health that did the checks for you.

ComicIronic commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+	var/no_drop = 0
+
+/obj/structure/block/attackby(var/obj/item/weapon/W,var/mob/user)
+	if(iswrench(W))
```
There's already a wrenchAnchor proc on /obj/, so you should use that. Same for the mannequin.

JMWTurner replied:
replaced.

ComicIronic commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+			"vox",
+			)
+
+		var/chosen_sculpture = input("Choose a sculpture type.", "[name]") as null|anything in available_sculptures
```
You need to repeat the Adjacency check after this input.

JMWTurner replied:
added.

ComicIronic commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+	overlays.len = 0
+	..()
+	if(destroyed)
+		overlays |= "mannequin_cover_broken"
```
Overlays should be using the image system.

ComicIronic commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1298 @@
+				Co_VIS_MSG = "{USER} install{s} the circuitboard into {HOLDER}.",
+				Co_AMOUNT = 1
+				),
+			Co_BACKSTEP = list(
```
You shouldn't declare a backstep for the first step, I don't know what kind of issues it could cause.

Exxion replied:
At least it's actually using construction datums
Also, optional steps when?

JMWTurner replied:
@ComicIronic Makes you wonder what's the point of reversible construction datums if there is no result for going that way.

ComicIronic replied:
You can go back and forth through the steps, that's the whole point. It'll be more useful after we (eventually) add multiple options per step for stuff like mechs and spacepods with custom plates.

Optional steps were giving me a lot of grief to implement, but I might go back to them.

JMWTurner commented:
revision 1
* Fixed a few issues, also added a few mapping_ vars to mannequin so mappers can set up ones already wearing clothes.

clusterfack commented:
It seems like a lot of this code is copypasted from other sources?

JMWTurner commented:
@clusterfack the update procs are inspired from human/update_icon.dm. The examine procs from human.dm. That's about it.

clusterfack commented:
I'm pretty concerned about the maintainability of that

JMWTurner commented:
I was concerned about making them as modular as possible. Currently you can easily add mannequins for other species, and it'll take their clothing restrictions into account. In fact I'm currently making a monkey mannequin child type and it's really easy with the framework I've made.

FudgePucker commented:
Ideas to be implemented later.
Cultists being able to possess them to turn them into simple mobs or ghost controlled.
Can't be used to summon.
Can't use guns.
Wood gets fucked over by burn damage
Marble has its limbs blown off by brute damage
Cyber gets disabled by stuns and destroyed by ions.

JMWTurner commented:
@FudgePucker Ideas are the last thing I need right now. I have hundreds of those myself. I'm trying really hard to not add too many features in this PR already.

ComicIronic commented:
I wish I could tag individual comments with WYCI.

PJB3005 commented:
If only....

FudgePucker commented:
That's why I said later.

I currently am looking forward to making a statue garden of marble solders in armor to guard my tomb

PJB3005 commented on code/game/objects/structures/displaycase.dm:
```diff
@@ -168,7 +168,7 @@
 	else
 		src.icon_state = "glassbox2[locked]"
-	overlays = 0
+	overlays.len = 0
```
This change does nothing and I wouldn't even be surprised if this made it _slower_.

JMWTurner replied:
https://github.com/d3athrow/vgstation13/pull/1711#discussion_r19302706

ZEFIOJHEZFJIOEZJHFI0ZEOHNFPOZEJFOIZEJN? FIKOZEHNFIVGKO NIK VNIOKZEAN FVGKE.SQZNIOJFVKEAQNIKO VAEOIH9FEAAZQESFCAZQSFCVAZQFVZA

PJB3005 replied:
Welp

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+
+
+/obj/structure/mannequin/Destroy()
+	new /obj/effect/decal/cleanable/dirt(loc)
```
I'm really against doing stuff like this in destroy, because it gets called from too many places.

Should an admin deleting a mannequin cause dirt to drop? No.

ComicIronic replied:
It should also be pooled.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+
+
+/obj/structure/mannequin/attack_hand(var/mob/user)
+	if(user.a_intent == I_HURT)
```
Maybe make this require the user to actually be on kicking intent?

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	spawn()
+		qdel (src)
+
+/datum/construction/reversible/mannequin
```
But why?

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+		C.req_access = const_holder.req_access
+		C.req_one_access = const_holder.req_one_access
+
+		spawn()
```
Why is this behind a `spawn()`?

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	const_holder.overlays.len = 0
+	const_holder.icon_state = "mannequin_cyber_human"
+	const_holder.overlays |= icon(const_holder.icon, "lightout")
+	spawn()
```
Same deal.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	desc = "Lots of work just to display some clothes."
+	icon = 'icons/obj/mannequin_64x64.dmi'
+	icon_state="mannequin_cyber_human_frame"
+	pixel_x = -16
```
Add the pixel multiplier, please.

ComicIronic replied:
Should this even be set in code? What's the reason to have manual pixel displacement from the start?

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	if(destroyed)
+		show_inv(user)
+	else if(locked)
+		if(user.a_intent == I_HURT)
```
Again, kick intent.

JMWTurner replied:
I'm adding a kick_act and replacing the "kick" in there with a "punch"

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+		overlays |= image(icon, "mannequin_cover")
+
+/obj/structure/mannequin/cyber/acidable()
+	return 0
```
Why?

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+			new_frame.construct = new /datum/construction/reversible/mannequin(new_frame)
+			qdel(src)
+
+	else if(istype(W, /obj/item/weapon/weldingtool) && (user.a_intent == I_HELP))
```
Redundant parentheses.

JMWTurner replied:
I'll keep those for legibility

PJB3005 replied:
...What?

JMWTurner replied:
Haven't you ever added parentheses around multiplications even if they aren't needed because it makes it easier to read?

PJB3005 replied:
That's arithmetic though, not logical expressions.

JMWTurner replied:
Here in this case you could say that "&&" is a "+" and "==" is a "x". Logical expression priorities work the same way that arithmetic ones do.

PJB3005 replied:
> Here in this case you could say that "&&" is a "+" and "==" is a "x"

Sorry but what? Now you're just making crap up for the sake of not having to change it.

> Logical expression priorities work the same way that arithmetic ones do.

... Only if you come up with dumb logic. It can be argued about the `+`/`&&` but `==`? Not really.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	..()
+	update_icon()
+
+/obj/structure/mannequin/cyber/Destroy()
```
Jesus christ this really isn't the point of `Destroy`.

Move all this spawning junk crap to the things that it logically makes sense for, not this.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	desc = "Holy shit."
+	icon = 'icons/obj/mannequin_64x64.dmi'
+	icon_state="mannequin_cyber_human"
+	pixel_x = -16
```
Pixel multiplier.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	icon = 'icons/obj/mannequin_64x64.dmi'
+	icon_state="mannequin_cyber_human"
+	pixel_x = -16
+	clothing_offset_x = 16
```
Pixel multiplier

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	icon_state="mannequin_cyber_human"
+	pixel_x = -16
+	clothing_offset_x = 16
+	clothing_offset_y = 7
```
Pixel multiplier.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+	time_to_sculpt = 100
+	result = "/obj/structure/mannequin/wood"
+
+/obj/structure/block/wood/Destroy()
```
You get the drill by now.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1386 @@
+		return wrenchAnchor(user, 50)
+	else if(istype(W, /obj/item/weapon/chisel))
+
+		var/list/available_sculptures = list(
```
Make this a static list.

JMWTurner replied:
Does it make a difference? Genuinely asking.

RemieRichards replied:
this list is always the same, so there's no point creating a new `list()` per call of the proc, making it static means the EXACT SAME LIST object is reused between calls, and yes it works inside of procs/verbs.

JMWTurner replied:
I see.

By the way @RemieRichards I've meant to ask, do you regularly check every coderbase's github? For a tg coder I've seen you here pretty often.

JMWTurner commented:
revision 2
* Added monkey mannequins
![monkey](https://cloud.githubusercontent.com/assets/21122521/18406870/99ba7f06-7703-11e6-9a43-141a825f05a4.png)

* Fixed various issues
* You can now kick mannequins

JMWTurner commented:
revision 3
* simple animals can attack mannequins as well now

clusterfack commented:
Yeah this doesn't look even remotely maintainable to me, it's all just hardcoded to EMULATE how other mobs do their icon code.

TheFriggenSawbones commented:
please fix this jmwturnerlink

DrCelt commented:
I can't support a PR with this much copy paste. It's impossible.

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -340,13 +351,18 @@
+
+/obj/structure/mannequin/proc/breakDown()
+	getFromPool(/obj/effect/decal/cleanable/dirt,loc)
+	qdel()
```
That should be `qdel(src)`

JMWTurner commented:
@clusterfack Then why don't you let me maintain it?

RemieRichards commented:
@JMWTurner this is absolutely not the cleanest or most modular way.
While ~strange~ sounding I think the optimal way would be a ```/mob/living/carbon/human/dummy/mannequin``` type (```/dummy``` for the invulnerability cos it's a dummy), this would have slightly altered hooks that allow it to act from inside an object, and be interacted with, and this human would sit INSIDE the mannequin object, and the mannequin object would just copy the overlays of the human (minus limb/damage/fire etc. overlays)

JMWTurner commented:
@RemieRichards So basically starting again from scratch?

PJB3005 commented:
Next up, people putting mannequins into sleepers and making them fall asleep.

It's the perfect plan.

RemieRichards commented:
@JMWTurner afraid so, I actually did all I said above myself for tg (sorry, i liked the idea!) but got super annoyed with our version of some of the inventory hooks.
It may work out better for you guys.

However one thing was definite, it was -infinitely- shorter code, cos it reused 99% of it from humans (without copypasta), so while it's "starting from scratch" it's not much work at all.

JMWTurner commented:
I'd really like to keep actual mobs out of that.

RemieRichards commented:
@JMWTurner I think it's pretty clear that the state of the PR as of now is completely unmergable, as a /tg/ maintainer I can say this wouldn't get in at all, and I know some /vg/ maintainers are just as anal as we are if not more so in places.

JMWTurner commented:
That's why I'm going to tweak the way overlays are added, but I'm not using your way.

Code is a wonderful thing, the same result can be achieved in a nigh inifinite number of ways. I'll find a way that both satisfies our maintainers and is original.

RemieRichards commented:
I don't think any re-implementation of an overlay system is good at all for this, but suit yourself I suppose, I just offered the most logical solution.

PJB3005 commented:
Yeah @RemieRichards, using a mob for this is way better.

Also Skyrim memes.

RemieRichards commented:
@PJB3005 skyrim memes? ;_: <i>I don't get the joooooooke~</i>

JMWTurner commented:
Mannequins in skyrims are actual mobs I believe.

PJB3005 commented:
@RemieRichards in Skyrim, mannequins (in the Hearthfire DLC) were basically static characters that couldn't do anything but get clothes equipped.

Bethesda fucked up and sometimes after reentering your house, the mannequins would begin walking around.

RemieRichards commented:
Haha wow, that's brilliant.

JMWTurner commented:
Wouldn't radstorms cause the mobs inside to mutate as well?

RemieRichards commented:
@JMWTurner ```/dummy``` comes built in with ```status_flags & GODMODE```, which should make it immune to most things.

JMWTurner commented:
It doesn't. I had horns grow on my head while testing stuff as a dummy during a radstorm.

RemieRichards commented:
Might want to set a new flag/correct that then, all work should be on a new ```/mannequin``` anyway so no real worries about altering existing functionality for dummies/humans.

JMWTurner commented:
Why isn't the current form "maintainable" ?

RemieRichards commented:
@JMWTurner to maintain any form of accurate recreation of how humans render things, both of these sections of code (the original, and your pasta) need to be updated.

if either section is bugged, they both need to be updated.
if either section gets refactored, they both need to be refactored.

it's just a standard practice, DRY, <B>D</B>on't <B>R</B>epeat <B>Y</B>ourself, it's one of the reasons we have functions and classes at all, it's all about code reuse without having to maintain 50 copies of something, you just alter the function/class.

PJB3005 commented:
@JMWTurner to be honest mannequins growing horns sounds funny enough to keep.

JMWTurner commented:
Out of question.

JMWTurner commented:
@RemieRichards So if I needed to make a mannequin mob. Do I also need to make a new species for every race of mannequins? How do you handle vox/monkey/fat mannequins on tg?

RemieRichards commented:
I mentioned I actually never finished it @JMWTurner because our inventory hooks are shitfuck and I was in no mood, I do still have the code around to pick it up later if I want.

a mannequin ```/species``` could be acoomplished in a similar way I did my ```/synth``` species, where the species holds a fake extra species datum, and all it's procs redirect to the fake one, so for the mannequin you'd change the fake one and it'd *look* like a Vox, while still being ```/datum/species/mannequin```

JMWTurner commented:
Fine. I give up.

TheFriggenSawbones commented:
PLEASE DONT GIVE UP JMWTURNER

ComicIronic commented:
Another way would be to just abstract the icon generation to a datum and then have both human and mannequin code use it, so you aren't hodge-podging together a solution with a mannequin mob.

JMWTurner commented:
That sounds reasonable. I'll re-open this PR whenever I manage to do that.

JMWTurner commented:
revision 7:
* copy&paste begone

Arthurdentist commented:
As a dummy abusing fuckwit here are some of the interesting things you can do with test dummies.

1. Give them viruses
2. Staff of Change them
3. Inject them with Advanced Slime Toxin (turning them into slimes)

Those are three things off the top of my head your dummy mannequin people would have to be made immune too.  There are probably a lot more things that I am not aware of or have forgotten.

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -472,399 +417,246 @@
+		else
+			return null
 
+/obj/structure/mannequin/proc/get_primitive_icons(var/slot)
```
This should surely inherit from the proc above somehow, as it's copy pasted from there.

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -472,399 +417,246 @@
 
-/obj/structure/mannequin/proc/update_icon_left_hand()
-	return update_icon_hand(GRASP_LEFT_HAND)
+/obj/structure/mannequin/proc/get_species_icons(var/slot)
```
Same with above.

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -472,399 +417,246 @@
 
-/obj/structure/mannequin/proc/update_icon_right_hand()
-	return update_icon_hand(GRASP_RIGHT_HAND)
+/obj/structure/mannequin/proc/get_fat_icons(var/slot)
```
Oh no not again

JMWTurner commented:
@Arthurdentist I'm not gonna use dummies. Too many factors, and too many restrictions.

JMWTurner commented:
revision 8
* even less copypasta

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+
+
+/obj/structure/mannequin/examine(mob/user)
+	..()
```
c o p y p a s t i n g 
is BAD

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+
+
+
+/obj/structure/mannequin/Topic(href, href_list)
```
This is copypasted too

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+			visible_message("\The [src] collapses.")
+			breakDown()
+
+/obj/structure/mannequin/cyber/attackby(var/obj/item/weapon/W,var/mob/user)
```
is this really all copypasted from the captain's laser holding thingy

JMWTurner replied:
It's not, there's no fingerprint indentification thing

DrCelt commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+			user.visible_message("<span class='danger'>[user.name] punches \the [src]!</span>", "<span class='danger'>You punch \the [src]!</span>", "You hear glass crack.")
+			getDamage(rand(1,7) * (user.get_strength() - 1))
+		else
+			to_chat(user,"<span class='notice'>You gently run your hands over \the [src] in appreciation of its contents.</span>")
```
it IS all copypasted from the captain's laser holding thingy

JMWTurner replied:
It's a fucking feedback message to player.

ComicIronic replied:
Celt, there's a difference between bad copypaste and code that exists twice. It's bad copypaste if they HAVE to be identical. In this case, it's just interaction. Could have been written from scratch.

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+	O.layer = FLOAT_LAYER
+	O.overlays.len = 0
+
+	update_icon_slot(O,SLOT_MANNEQUIN_ICLOTHING)
```
This should be a softcoded loop through clothing

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+		O.overlays += I
+
+/obj/structure/mannequin/proc/get_all_slot_icons()
+	all_slot_icons[SLOT_MANNEQUIN_ICLOTHING] = get_slot_icons(SLOT_MANNEQUIN_ICLOTHING)
```
Also should be a softcoded loop

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+	name = "monkey marble mannequin"
+	icon_state="mannequin_marble_monkey"
+	primitive = 1
+	clothing_offset_y = -5*PIXEL_MULTIPLIER
```
If monkeys and this both use this, this should be externally defined somewhere

JMWTurner replied:
They don't. Monkeys are sprited in the center of the tile so items in their hands (for instance) appear without any offset, just as if they were in human hands.

clusterfack commented on code/setup.dm:
```diff
@@ -243,6 +243,21 @@ var/MAX_EXPLOSION_RANGE = 14
+#define SLOT_MANNEQUIN_MASK			"mask"
+#define SLOT_MANNEQUIN_HEAD			"hat"
+#define SLOT_MANNEQUIN_BACK			"backpack"
+#define SLOT_MANNEQUIN_ID			"idcard"
```
Why aren't all these defines like ```#define SLOT_MANNEQUIN_ID SLOT_ID```?

JMWTurner replied:
I'll try, fingers crossed it doesn't break anything.

JMWTurner replied:
As I feared it didn't work. Because of all that "lists and numbers and indexes" problem

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+
+	var/inv_slot
+
+	switch(item_slot)
```
This would be unnecessary if the defines of these were equal to one another, why are they distinct?

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+/obj/structure/mannequin/proc/get_slot_icons(var/slot)
+	var/list/slotIcon = list()
+	switch(slot)
+		if(SLOT_MANNEQUIN_ICLOTHING)
```
It disturbs me that there isn't a polymorphic way to get these.

JMWTurner replied:
Ideally the human update_icon proc should be reworked so we could use the same proc for both mannequins and humans, but this is way outta my league.

JMWTurner replied:
I'll be waiting for someone to rework mob inventories before investing time in designing polymorphic procs that only mannequins would use anyway. I'll put this PR on hold if this is what it takes.

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+			return null
+
+/obj/structure/mannequin/proc/checkMappingWear()
+	if(mapping_uniform)
```
I feel like there's a way this COULD be fully softcoded. But at the least you could simply do...
var/list/stuff = list(slot_mannequin_iclothing = mapping_uniform, slot_mannequin_feet = mapping_shoes,...)

and then simply loop through that with an EXTREMELY short loop with almost zero copy pasting.

JMWTurner replied:
Just like I did with the icons updates then. I'm seeing a pattern here.

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+
+
+/obj/structure/mannequin/attack_hand(var/mob/living/user)
+	if(user.a_intent == I_HURT)
```
I think you need to call ..() or it doesn't get an attack delay? Something like that

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+
+/obj/structure/mannequin/MouseDrop(var/mob/M)
+	..()
+	if(M != usr)
```
This probably needs a lot more sanity, mousedrop/click other procs like that are the least sane you could possibly be

clusterfack commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1187 @@
+/obj/structure/mannequin/proc/get_slot_icons(var/slot)
+	var/list/slotIcon = list()
+	switch(slot)
+		if(SLOT_MANNEQUIN_ICLOTHING)
```
It's not **as** hard as it sounds. Just requires a bit of OOP thinking.

clusterfack replied:
Maintainability wise, this is my only concern left. If this gets resolved I think this PR will be fairly ready to merge.

clusterfack commented on code/setup.dm:
```diff
@@ -243,6 +243,21 @@ var/MAX_EXPLOSION_RANGE = 14
+#define SLOT_MANNEQUIN_MASK			"mask"
+#define SLOT_MANNEQUIN_HEAD			"hat"
+#define SLOT_MANNEQUIN_BACK			"backpack"
+#define SLOT_MANNEQUIN_ID			"idcard"
```
To be honest if these don't have any reason to be distinct you could just remove them entirely and just use the normal slot defines

JMWTurner commented:
revision 9
* yet less copypaste

Intigracy commented:
oh lawdy 135 comments

JMWTurner commented:
Fixed conflicts

JMWTurner commented:
How about now @clusterfack ?

clusterfack commented:
Yeah I think it's fine, I'd like another collaborator to look at it as well to verify it

PJB3005 commented on code/game/objects/structures/mannequin.dm:
```diff
@@ -0,0 +1,1096 @@
+
+	for(var/obj/item/I in held_items)
+		if(I.blood_DNA && I.blood_DNA.len)
+			msg += "<span class='warning'>I's holding [bicon(I)] [I.gender==PLURAL?"some":"a"] blood-stained [I.name] in its [get_index_limb_name(is_holding_item(I))]!</span>\n"
```
I's -> It's

gbasood replied:
Fix this.

JMWTurner commented:
@gbasood fix'ed

gbasood commented:
@JMWTurner What about the line comment I mentioned?
https://github.com/d3athrow/vgstation13/pull/11821#discussion_r78853081

JMWTurner commented:
@gbasood commented as you wished. I'm fairly sure I already explained in this PR why I had to do that.

gbasood commented:
@JMWTurner Wait, hold on, are we talking about the same one?
Because I don't see a comment still and I don't see anything else about it
![](http://i.imgur.com/Kdy8dwY.png)

JMWTurner commented:
@gbasood It is the same line. It used to be at line 398 when cluster commented but it is now at 366.

The github code comments don't update anymore it seems. Look at my last commit.

Intigracy commented:
They only update if the line you comment on changes.

Probe1 commented on code/datums/supplypacks.dm:
```diff
@@ -290,6 +290,13 @@ var/list/all_supply_groups = list("Supplies","Clothing","Security","Hospitality"
 	containername = "Arts and Crafts crate"
 	group = "Supplies"
 
+/datum/supply_packs/artscrafts
```
https://github.com/d3athrow/vgstation13/issues/12718
